### PR TITLE
Change Address alias to Vibe URL

### PR DESCRIPTION
**Pull request by mkykadir**
Part of #2566 

- Remove `string` alias of `Address`
- Extend Vibe's `URL` with serialization under `Address` structure

It might be better to change URL expecting inputs of `Config` from `string` to `Address`. It is not a problem for mapping types (as in `Proxy` config) but sequence types cannot be parsed with string inputs into `URL` struct yet. This way we would be able to throw error at configuration parsing when user is provided invalid URL from config file.

### CODE REVIEW

File: source/agora/common/BanManager.d [hunk 1]
```diff
@@ -381,50 +381,54 @@ unittest
         public override void load () { }
     }
 
+    myAddress node1 = myAddress("agora://node-1");
+    myAddress node2 = myAddress("agora://node-2");
+    myAddress whitelistNode = myAddress("agora://whitelist-node");
```
Comment (Geod24) [inline source/agora/common/BanManager.d]:
```suggestion
    const node1 = myAddress("agora://node-1");
    const node2 = myAddress("agora://node-2");
    const whitelistNode = myAddress("agora://whitelist-node");
```

Can you create a commit with those changes (using `Address` instead of `myAddress` of course) and submit it in its own PR ? That should be a no-brainer to merge.

File: source/agora/flash/Config.d [hunk 2]
```diff
@@ -50,7 +52,7 @@ public struct FlashConfig
     public ushort control_port = 0xB0C;
 
     /// Address to the listener which will receive payment / update notifications
-    public string listener_address;
+    public @Optional string listener_address;
```
Comment (Geod24) [inline source/agora/flash/Config.d]: ?

Comment (mkykadir) [reply source/agora/flash/Config.d]: Stale change, will be reverted

**Comment by Geod24** [comment]
```
node-7_1  | 2021-10-25T09:34:05.844974693Z [main(m8Tl) INF] Listening for requests on http://0.0.0.0:7826/
node-7_1  | 2021-10-25T09:34:05.851552114Z Timer callback failed: Empty URL.
node-7_1  | 2021-10-25T09:34:06.085775498Z Full exception: object.Exception@submodules/vibe.d/inet/vibe/inet/url.d(145): Empty URL.
node-7_1  | 2021-10-25T09:34:06.085810700Z ----------------
node-7_1  | 2021-10-25T09:34:06.085815001Z /usr/include/d/std/exception.d:513 pure @safe void std.exception.bailOut!(Exception).bailOut(immutable(char)[], ulong, scope const(char)[]) [0x5622c5a56010]
node-7_1  | 2021-10-25T09:34:06.085818501Z /usr/include/d/std/exception.d:433 pure @safe bool std.exception.enforce!().enforce!(bool).enforce(bool, lazy const(char)[], immutable(char)[], ulong) [0x5622c5a55f90]
node-7_1  | 2021-10-25T09:34:06.085821901Z submodules/vibe.d/inet/vibe/inet/url.d:142 ref @safe vibe.inet.url.URL vibe.inet.url.URL.__ctor(immutable(char)[]) [0x5622c64d4c80]
node-7_1  | 2021-10-25T09:34:06.085825201Z source/agora/common/Types.d:51 ref @safe agora.common.Types.myAddress agora.common.Types.myAddress.__ctor(immutable(char)[]) [0x5622c5d1ecb0]
node-7_1  | 2021-10-25T09:34:06.085828202Z source/agora/network/Manager.d:667 agora.common.Task.ITimer agora.network.Manager.NetworkManager.startPeriodicNameRegistration() [0x5622c5d138e0]
```

File: source/agora/network/Manager.d [hunk 3]
```diff
@@ -667,7 +667,7 @@ public class NetworkManager
     public ITimer startPeriodicNameRegistration ()
     {
         this.registry_client = this.getNameRegistryClient(
-            this.validator_config.registry_address, 2.seconds);
+            myAddress(this.validator_config.registry_address), 2.seconds);
```
Comment (Geod24) [inline source/agora/network/Manager.d]: So this throws because `registry_address` is `null`.

Comment (mkykadir) [reply source/agora/network/Manager.d]: Yes, I've updated check wrong. It was checking in `getNameRegistryClient`, will update

File: source/agora/common/Set.d [hunk 4]
```diff
@@ -66,21 +66,23 @@ public struct Set (T)
     public void toString (scope void delegate(const(char)[]) @safe sink) const
     {
         import std.format : formattedWrite;
+        import std.conv : to;
 
         formattedWrite(sink, "[%(%s, %)]", this._set.byKeyValue().filter!(
-            kv => kv.value).map!(kv => kv.key));
+            kv => kv.value).map!(kv => kv.key.to!string));
```
Comment (Geod24) [inline source/agora/common/Set.d]: Wouldn't:
```suggestion
        formattedWrite(sink, `[%("%s", %)]`, this._set.byKeyValue().filter!(
            kv => kv.value).map!(kv => kv.key));
```

Be better as it would not allocate an intermediate string ?

Comment (mkykadir) [reply source/agora/common/Set.d]: This would result in double quotation marks in `Set!string`

Comment (Geod24) [reply source/agora/common/Set.d]: Is it the only case ?

File: source/agora/node/Config.d [hunk 5]
```diff
@@ -630,7 +629,7 @@ network:
   - http://192.168.0.42:2826
 `;
         auto conf = parseConfigString!Config(conf_str, "/dev/null");
-        assert(conf.network == [ `http://192.168.0.42:2826` ]);
+        assert(conf.network == [ "http://192.168.0.42:2826" ]);
```
Comment (Geod24) [inline source/agora/node/Config.d]: Why the change ?

File: source/agora/network/Manager.d [hunk 6]
```diff
@@ -237,7 +237,8 @@ public class NetworkManager
             const is_validator = key != PublicKey.init;
             if (is_validator)
             {
-                if (address.length && key == this.outer.validator_config.key_pair.address)
+                if (address != Address.init
```
Comment (Geod24) [inline source/agora/network/Manager.d]: Prefer `!is` for comparison default-initialized `struct`.

File: source/agora/node/Runner.d [hunk 7]
```diff
@@ -139,7 +140,10 @@ public Listeners runNode (Config config)
     }
 
     bool delegate (in NetworkAddress address) @safe nothrow isBannedDg = (in address) @safe nothrow {
-        return result.node.getBanManager().isBanned(address.toAddressString());
+        try
+            return result.node.getBanManager().isBanned(Address("agora://" ~ address.toAddressString()));
```
Comment (Geod24) [inline source/agora/node/Runner.d:144]: This should just use the `host`.

File: source/agora/test/Base.d [hunk 8]
```diff
@@ -2271,7 +2273,7 @@ public APIManager makeTestNetwork (APIManager : TestAPIManager = TestAPIManager)
       node : makeNodeConfig(),
       interfaces: [ makeInterfaceConfig("name.registry") ],
       consensus: makeConsensusConfig(),
-      network : net.nodes.map!(pair => pair.address).array.dup,
+      network : cast(immutable) net.nodes.map!(pair => "http://"~pair.address).array.dup,
```
Comment (Geod24) [inline source/agora/test/Base.d]:
```suggestion
      network : net.nodes.map!(pair => "http://"~pair.address).array.idup,
```

Comment (Geod24) [reply source/agora/test/Base.d]: Or:
```suggestion
      network : cast(immutable) net.nodes.map!(pair => "http://"~pair.address).array,
```

File: source/agora/test/Base.d [hunk 9]
```diff
@@ -532,7 +532,7 @@ public class TestBanManager : BanManager
 public struct NodePair
 {
     ///
-    public Address address;
+    public string address;
```
Comment (Geod24) [inline source/agora/test/Base.d:535]: Why this change ?

File: source/agora/test/Base.d [hunk 10]
```diff
@@ -2199,24 +2201,24 @@ public APIManager makeTestNetwork (APIManager : TestAPIManager = TestAPIManager)
...
         };
 
         return conf;
     }
 
-    string validatorAddress (size_t idx, KeyPair key)
+    Address validatorAddress (size_t idx, KeyPair key)
     {
-        return format("Validator-%s.%s.localrest", idx, key.address);
+        return Address(format("http://Validator-%s.%s.localrest", idx, key.address));
```
Comment (Geod24) [inline source/agora/test/Base.d:2216]: Since Localrest uses the binary protocol, I think it would make more sense to use `agora://` here. Or its own `localrest://` scheme. But the current approach works, so let's keep that for another day.

**Review by Geod24 — APPROVED** [review]
Broadly LGTM. Just a few minor points.

File: source/agora/test/Base.d (hunk 8 above)
Comment (Geod24) [reply source/agora/test/Base.d]: Note that this applies to other instances in this file as well.

File: source/agora/test/Base.d (hunk 9 above)
Comment (mkykadir) [reply source/agora/test/Base.d:535]: `NodePair`s are [generated from](https://github.com/bosagora/agora/blob/ef4662921f3aaaae2d733ab3055b899d2f690829/source/agora/test/Base.d#L875) interfaces configuration only with addresses (without schema or port information)

**Review by Geod24 — APPROVED** [review]
